### PR TITLE
[hw-agnostic] Adding embedding layer and lm_head

### PR DESCRIPTION
**Pull request by bohnstingl**
## Purpose

Extends the hardware-agnostic layer path in the Transformers modeling backend to
cover the token-embedding boundary: input embedding (`VocabParallelEmbedding`),
LM head (`ParallelLMHead`), and the logits projection (`LogitsProcessor`).

When `VLLM_USE_HW_AGNOSTIC` is unset (the default), behavior is unchanged: every
site resolves to the same vLLM class it used before.

## Test Plan

I ran an `lm_eval` with the Granite3.3-8B model, to demonstrate that the new additional layers do not introduce numerical issues.

```
export VLLM_USE_HW_AGNOSTIC=1
lm_eval --model vllm \
  --model_args "pretrained=ibm-granite/granite-3.3-8b-instruct,model_impl=transformers,tensor_parallel_size=1,gpu_memory_utilization=0.85,max_model_len=4096,enforce_eager=True" \
  --tasks gsm8k \
  --num_fewshot 5 \
  --batch_size auto \
  --output_path /block/boh/vllm_dev/claude/granite_eval_logs/flag_on
```
In addition, I also added some new, specific tests for the layers.

## Test Result

lm_eval results:

Metric | OFF | ON | Δ
-- | -- | -- | --
flexible-extract | 0.7240 ± 0.0123 | 0.7225 ± 0.0123 | −0.0015
strict-match | 0.6710 ± 0.0129 | 0.6664 ± 0.0130 | −0.0046

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [X] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [X] The test plan, such as providing test command.
- [X] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

cc @tdoublep @hmellor @zou3519

### CODE REVIEW

File: vllm/model_executor/models/transformers/causal.py [hunk 1]
```diff
@@ -47,19 +50,23 @@ def __init__(self, *, vllm_config: "VllmConfig", prefix: str = ""):
...
                 self.text_config.vocab_size,
                 self.text_config.hidden_size,
                 quant_config=self.quant_config,
                 prefix=maybe_prefix(prefix, "lm_head"),
             )
             if tie_word_embeddings:
+                vpe_classes = (
+                    get_vocab_parallel_embedding_cls(),
+                    VllmVocabParallelEmbedding,
+                )
```
Comment (hmellor) [inline vllm/model_executor/models/transformers/causal.py]: Why do we need to check both?

Comment (bohnstingl) [reply vllm/model_executor/models/transformers/causal.py]: We don't actually. This was just a defensive check. I reverted this in my latest commit.

File: vllm/model_executor/models/transformers/causal.py [hunk 2]
```diff
@@ -47,19 +47,20 @@ def __init__(self, *, vllm_config: "VllmConfig", prefix: str = ""):
...
                 self.text_config.vocab_size,
                 self.text_config.hidden_size,
                 quant_config=self.quant_config,
                 prefix=maybe_prefix(prefix, "lm_head"),
             )
             if tie_word_embeddings:
+                vpe_classes = get_vocab_parallel_embedding_cls()
                 for module in self.model.get_input_embeddings().modules():
-                    if isinstance(module, VocabParallelEmbedding):
+                    if isinstance(module, vpe_classes):
```
Comment (hmellor) [inline vllm/model_executor/models/transformers/causal.py]: Could we do this?

```suggestion
for module in self.model.get_input_embeddings().modules():
                    if isinstance(module, get_vocab_parallel_embedding_cls()):
```

Comment (bohnstingl) [reply vllm/model_executor/models/transformers/causal.py]: @hmellor I just did that, see above

Comment (hmellor) [reply vllm/model_executor/models/transformers/causal.py]: The current state still writes `get_vocab_parallel_embedding_cls()` to `vpe_classes`

Comment (bohnstingl) [reply vllm/model_executor/models/transformers/causal.py]: This is intentional, because I wanted to avoid re-evaluating `get_vocab_parallel_embedding_cls()` in the loop, but if you are in favor of that, I can do it?

Comment (hmellor) [reply vllm/model_executor/models/transformers/causal.py]: Oh I see, I missed that
